### PR TITLE
add ffmpeg decoder to extra/libs

### PR DESCRIPTION
**Pull request by FASTSHIFT**
### Description of the feature or fix

Add FFmpeg support, which can decode almost all pictures and videos, including resources on the Internet.

### Checkpoints
- [x] Follow the [styling guide](https://github.com/lvgl/lvgl/blob/master/docs/CODING_STYLE.md)
- [ ] Update [CHANGELOG.md](https://github.com/lvgl/lvgl/blob/master/docs/CHANGELOG.md)
- [x] Update the documentation

### CODE REVIEW

**Comment by FASTSHIFT** [comment]
Example of use:

image
```C
static lv_obj_t *ffmpeg_img_test(const char *path)
{
  lv_obj_t *img = lv_img_create(lv_scr_act());

  lv_img_set_src(img, path);
  return img;
}

static lv_obj_t *ffmpeg_video_test(const char *path)
{
  lv_obj_t *player = lv_ffmpeg_player_create(lv_scr_act());

  lv_ffmpeg_player_set_src(player, path);
  lv_ffmpeg_player_set_auto_restart(player, true);
  lv_ffmpeg_player_set_cmd(player, LV_FFMPEG_PLAYER_CMD_START);
  return player;
}

int main(int argc, char **argv)
{
  if (argc != 2)
  {
    printf("usage: lv_ffmpeg input_file\n");
    return -1;
  }

  const char *path = argv[1];

  if (access(path, F_OK) < 0)
  {
    printf("file: %s was not found!\n", path);
    return -1;
  }

  /* LVGL initialization */

  lv_init();

  hal_init();

  lv_obj_t* img;

  if (lv_ffmpeg_get_frame_num(path) > 1)
  {
    img = ffmpeg_video_test(path);
  }
  else
  {
    img = ffmpeg_img_test(path);
  }

  lv_obj_center(img);

  while (1)
  {
    lv_timer_handler();
    usleep(5000);
  }

  return 0;
}

```

File: lv_conf_template.h [hunk 1]
```diff
@@ -581,6 +581,13 @@ e.g. "stm32f769xx.h" or "stm32f429xx.h"*/
 /*Rlottie library*/
 #define LV_USE_RLOTTIE 0
 
+/*FFmpeg library*/
+#define LV_USE_FFMPEG  0
```
Comment (xiaoxiang781216) [inline lv_conf_template.h:586]: update Kconfig too

Comment (FASTSHIFT) [reply lv_conf_template.h:586]: Done

File: lv_conf_template.h [hunk 2]
```diff
@@ -581,6 +581,13 @@ e.g. "stm32f769xx.h" or "stm32f429xx.h"*/
 /*Rlottie library*/
 #define LV_USE_RLOTTIE 0
 
+/*FFmpeg library*/
```
Comment (kisvegabor) [inline lv_conf_template.h]:
```suggestion
/*FFmpeg library for image decoding and playing videos
 *Supports all major image formats so do not enable other image decoder with it*/
```

**Review by kisvegabor — COMMENTED** [review]
Wow, it's amazing!

A couple of questions and notes:
1. What will be the size of the video player? Is it set automatically to the source video's size? Or will scale the video?
2.  How should the `path` look like? Normally e.g. "../foo.jpg" or with an LVGL driver letter, e.g. "S:foo.jpg"?
3. Could you add a small example image and a very short video next to the examples and load them? 
4. Please add a new file in the [docs](https://github.com/lvgl/lvgl/tree/master/docs/libs) too. (it also needs to be added to [index.md](https://github.com/lvgl/lvgl/blob/master/docs/libs/index.md))

**Comment by FASTSHIFT** [comment]
@kisvegabor 
1.Video will only buffer one frame, read and refresh from the file regularly.
2.The path uses the form of `"../foo.jpg"`, because the path is passed directly to FFmpeg without going through the file system interface of lvgl.The path can be a URL, because FFmpeg supports getting pictures from the Internet.

File: lv_conf_template.h (hunk 2 above)
Comment (FASTSHIFT) [reply lv_conf_template.h]: Done

**Comment by kisvegabor** [comment]
> 1.Video will only buffer one frame, read and refresh from the file regularly.

I meant the width and height :slightly_smiling_face:

**Comment by FASTSHIFT** [comment]
> > 1.Video will only buffer one frame, read and refresh from the file regularly.
> 
> I meant the width and height 🙂

By default, the width and height of the video player are the same as the original size of the video. You can also use `lv_img_set_zoom` to zoom the size.

**Comment by kisvegabor** [comment]
> By default, the width and height of the video player are the same as the original size of the video. You can also use lv_img_set_zoom to zoom the size.

Great! 
When the docs is updated I'll try it out by following the instructions from there.

**Comment by kisvegabor** [comment]
Thanks! I've just tested it and worked well, except that I needed to link `phtread` too. 

Two minor things: 
- Please create separate examples for the image and the video. It's because the online examples are shown on 320x240 resolution.
- Use [this technique](https://github.com/lvgl/lvgl/blob/master/examples/libs/rlottie/lv_example_rlottie_1.c#L15-L23) on the examples else the docs generator will fail.

**Review by embeddedt — COMMENTED** [review]
It might be best to pick a different video (e.g. one of the demos @kisvegabor uploaded to the LVGL YouTube channel) as I think this song is copyrighted.

**Comment by kisvegabor** [comment]
> as I think this song is copyrighted.

I was also thinking about it but forgot to mention it.  
What about this? https://www.videezy.com/abstract/44864-silhouettes-of-birds-over-the-sunset
We only need an attribution.

File: examples/libs/ffmpeg/lv_example_ffmpeg_2.c [hunk 3]
```diff
@@ -0,0 +1,28 @@
+#include "../../lv_examples.h"
+#if LV_USE_FFMPEG && LV_BUILD_EXAMPLES
+
+/**
+ * Open a video from a file
+ */
+void lv_example_ffmpeg_2(void)
+{
+    lv_obj_t * player = lv_ffmpeg_player_create(lv_scr_act());
```
Comment (kisvegabor) [inline examples/libs/ffmpeg/lv_example_ffmpeg_2.c:11]:
```suggestion
    /*birds.mp4 is downloaded from http://www.videezy.com (Free Stock Footage by Videezy!)
     *https://www.videezy.com/abstract/44864-silhouettes-of-birds-over-the-sunset*/
    lv_obj_t * player = lv_ffmpeg_player_create(lv_scr_act());
```

**Comment by kisvegabor** [comment]
I've just added the attribution in the video example.

All looks good to me, merging. Thank you very much for this super useful feature.

**Comment by FASTSHIFT** [comment]
@kisvegabor 
It seems that the document on the web page will not wrap.
image

**Comment by kisvegabor** [comment]
I've updated the CSS. 
image

**Comment by suijunqiang** [comment]
> Thanks! I've just tested it and worked well, except that I needed to link `phtread` too.
> 
> Two minor things:
> 
> * Please create separate examples for the image and the video. It's because the online examples are shown on 320x240 resolution.
> * Use [this technique](https://github.com/lvgl/lvgl/blob/master/examples/libs/rlottie/lv_example_rlottie_1.c#L15-L23) on the examples else the docs generator will fail.

Hello, @kisvegabor 
First of all, really thanks for your great contributions. 
I am trying to run the LVGL ffmpeg demo, but get ld error about can not find APIs in the FFMPEG libs, and I am using the [lv_port_win_vscode_master](https://gitee.com/yangwuju/lv_port_win_vscode-master.git) all example are running well but the ffmpeg.

so would you please give me some supports?

Thanks & Best Regards
Sui

**Comment by suijunqiang** [comment]
> > Thanks! I've just tested it and worked well, except that I needed to link `phtread` too.
> > Two minor things:
> > 
> > * Please create separate examples for the image and the video. It's because the online examples are shown on 320x240 resolution.
> > * Use [this technique](https://github.com/lvgl/lvgl/blob/master/examples/libs/rlottie/lv_example_rlottie_1.c#L15-L23) on the examples else the docs generator will fail.
> 
> Hello, @kisvegabor First of all, really thanks for your great contributions. I am trying to run the LVGL ffmpeg demo, but get ld error about can not find APIs in the FFMPEG libs, and I am using the [lv_port_win_vscode_master](https://gitee.com/yangwuju/lv_port_win_vscode-master.git) all example are running well but the ffmpeg.
> 
> so would you please give me some supports?
> 
> Thanks & Best Regards Sui

image

**Comment by kisvegabor** [comment]
Have you installed ffmpeg and added the it to your linker settings?

**Comment by suijunqiang** [comment]
> Have you installed ffmpeg and added the it to your linker settings?

Yeah, I have tried again, last issue gone.

but could not ld functions in ffmpeg .a when build.
errors as below:
/ld.exe: CMakeFiles\LVGL.dir/objects.a(lv_ffmpeg.c.obj): in function `av_make_error_string':
[build] D:/src/lv_port_win_vscode-master/lvgl/src/libs/ffmpeg/lv_ffmpeg.c:113:(.text+0x25): undefined reference to `av_strerror'
[build] C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\LVGL.dir/objects.a(lv_ffmpeg.c.obj): in function `lv_ffmpeg_init':

and also I have copied libavcodec.a/libavutil.a/libswscale.a/libavformat.a into lvgl/ffmpeg root folder and met same error.

the GIF reference as below:
ffmpeg

**Comment by suijunqiang** [comment]
> Have you installed ffmpeg and added the it to your linker settings?

image

**Comment by kisvegabor** [comment]
E.g. for `./ffmpeg/libavcodec.a` you should pass the library path as a linker flag too. E.g. `-Lpath/to/libs`

**Pull request merged**